File: 21_collision_ahead.cpp
// this is for emacs file handling -*- mode: c++; indent-tabs-mode: nil -*-
// -- BEGIN LICENSE BLOCK ----------------------------------------------
// -- END LICENSE BLOCK ------------------------------------------------
//----------------------------------------------------------------------
/*!\file
 *
 * \author  Christoph Rist <rist.christoph@gmail.com>
 * \date    2016-12-25
 *
 */
//----------------------------------------------------------------------
/***********************************
 * IEEE SB Passau Adventskalender
 * Problem 21
 */

#include <memory>
#include <algorithm>
#include <iostream>
#include <iomanip>
#include <vector>

//#define TEST
#ifdef TEST
#include "io_redirect.h"
#endif

using namespace std;

struct AABBRectangle
{
  AABBRectangle(double left, double right, double bottom, double top)
  : left(left), right(right), bottom(bottom), top(top)
  {}
  double left, right, bottom, top;

  bool collision(const AABBRectangle& other)
  {
    return bottom < other.top && top > other.bottom && left < other.right && right > other.left;
  }
};


int main(int argc, char* argv[])
{
#ifdef TEST
  //////////////////////// INPUT/OUTPUT //////////////////////////
  if (!redirect_io(argc, argv))
    return 0;
#endif

  uint32_t nb_rect;
  cin >> nb_rect;

  vector<AABBRectangle> rects;
  rects.reserve(nb_rect);

  for (uint32_t i = 0; i < nb_rect; ++i)
  {
    double left, right, bottom, top;
    cin >> left >> right >> bottom >> top;
    rects.emplace_back(left, right, bottom, top);
  }

  for (uint32_t i = 0; i < nb_rect; ++i)
  {
    cout <<i <<" with [";
    bool first = true;
    for (uint32_t j = 0; j < nb_rect; ++j)
    {
      if (j == i)
        continue;
      if (rects[i].collision(rects[j]))
      {
        if (first)
          first = false;
        else
        {
          cout <<", ";
        }
        cout <<j;
      }
    }
    cout <<"]" <<endl;
  }

#ifdef TEST
  cleanup_io();
#endif

  return 0;
}

File: 21_collision_ahead.py
#!/usr/bin/env python3
import sys

__author__ = "Christoph Rist"
__license__ = "MIT"

DEBUG = False


class AABBRectangle:
    def __init__(self, left, right, bottom, top):
        self.__dict__.update(locals())

    def collision(self, other):
        return self.bottom < other.top and \
                self.top > other.bottom and \
                self.left < other.right and \
                self.right > other.left


def main():
    if DEBUG:
        sys.stdin = open("samples/21.1_input.txt")

    nb_rectangles = int(input())

    if nb_rectangles > 65535:
        print("lll")
        return

    rect = []
    for n in range(nb_rectangles):
        rect.append(AABBRectangle(*filter(None, map(float, input().split(" ")))))

    collisions = [[] for _ in range(nb_rectangles)]
    for i in range(nb_rectangles):
        for j in range(i + 1, nb_rectangles):
            if rect[i].collision(rect[j]):
                collisions[i].append(j)
                collisions[j].append(i)

    if nb_rectangles > 100000:
        print("ll")
        return

    for i in range(nb_rectangles):
        print("{} with {}".format(i, collisions[i]))


if __name__ == "__main__":
    main()

